src/lib.rs

```diff
@@ -104,6 +107,124 @@ impl Invoice {
     fn version_in_range(&self, requirement: &str) -> bool {
         version_compare(self.bindle.id.version(), requirement)
     }
+
+    fn cleartext(&self, by: String, role: SignatureRole) -> String {
+        let id = self.bindle.id.clone();
+        let mut buf = vec![
+            by.clone(),
+            id.name().to_owned(),
+            id.version_string(),
+            format!("{:?}", role).to_lowercase(),
+            '~'.to_string(),
+        ];
+
+        // Add bindles
+        self.parcel.as_ref().map(|list| {
+            list.iter().for_each(|p| {
+                buf.push(p.label.sha256.clone());
+            })
+        });
+
+        buf.join("\n")
+    }
+
+    /// Sign the parcels on the current package.
+    ///
+    /// Note that this signature will be invalidated if any parcels are
+    /// added after this signature.
+    ///
+    /// In the current version of the spec, a signature is generated by combining the
+    /// signer's ID, the invoice version, and a list of parcels, and then performing
+    /// a cryptographic signature on those fields. The result is then stored in
+    /// a `[[siganture]]` block on the invoice. Multiple signatures can be attached
+    /// to any invoice.
+    pub fn sign(
```

Comment:
Non-blocking super nit: With as big as invoice is getting (and its associated types like `SignatureError`, it might be worth moving invoice to its own module like `Id` and then reexporting it at the top level